src/Config/ConfigProcessor.php

```diff
@@ -0,0 +1,156 @@
+     *
+     * @param ConfigLoaderInterface $loader
+     */
+    public function extend(ConfigLoaderInterface $loader)
+    {
+        return $this->add($loader->export(), $loader->getSourceName());
+    }
+
+    /**
+     * Extend the configuration to be processed with
+     * the provided nested array.
+     *
+     * @param array $data
+     * @param string $source
+     */
+    public function add($data, $source = '')
+    {
+        if (empty($source)) {
+            $this->unprocessedConfig[] = $data;
+            return $this;
+        }
+        $this->unprocessedConfig[$source] = $data;
+        return $this;
+    }
+
+    /**
+     * Process all of the configuration that has been collected,
+     * and return a nested array.
+     *
+     * @return array
+     */
+    public function export()
+    {
+        if (!empty($this->unprocessedConfig)) {
+            $this->processedConfig = $this->process(
+                $this->processedConfig,
+                $this->fetchUnprocessed()
+            );
+        }
+        return $this->processedConfig;
+    }
+
+    /**
+     * Get the configuration to be processed, and clear out the
+     * 'unprocessed' list.
+     *
+     * @return array
+     */
+    protected function fetchUnprocessed()
+    {
+        $toBeProcessed = $this->unprocessedConfig;
+        $this->unprocessedConfig = [];
+        return $toBeProcessed;
+    }
+
+    /**
+     * Use a map-reduce to evaluate the items to be processed,
+     * and merge them into the processed array.
+     *
+     * @param array $processed
+     * @param array $toBeProcessed
+     * @return array
+     */
+    protected function process($processed, $toBeProcessed)
+    {
+        $toBeReduced = array_map([$this, 'preprocess'], $toBeProcessed);
+        return array_reduce($toBeReduced, [$this, 'reduceOne'], $processed);
+    }
+
+    /**
+     * Process a single configuration file from the 'to be processed'
+     * list. By default this is a no-op. Override this method to
+     * provide any desired configuration preprocessing, e.g. dot-notation
+     * expansion of the configuration keys, etc.
+     *
+     * @param array $config
+     * @return array
+     */
+    protected function preprocess($config)
+    {
+        return $config;
+    }
+
+    /**
+     * Evaluate one item in the 'to be evaluated' list, and then
+     * merge it into the processed configuration (the 'carry').
+     *
+     * @param array $processed
+     * @param array $config
+     * @return array
+     */
+    protected function reduceOne($processed, $config)
+    {
+        $evaluated = $this->evaluate($processed, $config);
+        return static::arrayMergeRecursiveDistinct($processed, $evaluated);
+    }
+
+    /**
+     * Evaluate one configuration item.
+     *
+     * @param array $processed
+     * @param array $config
+     * @return array
+     */
+    protected function evaluate($processed, $config)
+    {
+        return Expander::expandArrayProperties(
```

Comment:
There's an issue here. This takes _new_ config and expands it's properties using values from the already processed config. That's good.

But, it's possible that the already processed config has unexpanded placeholders for which the new config provides values. In that case, we must re-expand the already processed config. Perhaps this could be done once against the entire data array just prior to export?